### ISSUE TITLE
Deadlock/crash on checkpointing many Functions on ExtrudedMesh

### ISSUE THREAD

**Issue by dham** [firedrakeproject/firedrake#2528]
The following code deadlocks on two processes after 183 iterations of the loop. Strangely, deleting one A from the name under which the Function is saved causes the crash whose output is given below. This appears to be another case of an attribute growing with the number of functions saved.

This test was conducted on Firedrake master with PETSc on the `JDBetteridge/gc_fix_hdf5_plex` branch (revision `1f74f506d8`).

```python
"""
   Parallel Checkpointing Test
"""
from firedrake import *
import os

try:
    os.remove("tmp.h5")
except FileNotFoundError:
    pass

# The deadlock appears only in case of ExtrudedMesh
use_extruded_mesh = True

if use_extruded_mesh:
    # 1D base mesh
    mesh1d = IntervalMesh(15, length_or_left=0.0, right=1.0)

    # extruding the base mesh "mesh1d" in the third dimension
    mesh = ExtrudedMesh(mesh=mesh1d, layers=15, layer_height=1./15,
                        extrusion_type='uniform'
                        )
else:
    # extruding the base mesh "mesh1d" in the third dimension
    mesh = UnitSquareMesh(nx=15, ny=15,
                          quadrilateral=True, name="UnitSquareMesh")

Q = FunctionSpace(mesh, "CG", 1)  # Temperature function space (scalar)
T = Function(Q)

with CheckpointFile("foo.h5", "a") as outfile:
    for timestep in ProgressBar().iter(range(0, 200)):
        outfile.save_function(T, name=f"AAAAAAAAAAA_{timestep}")
```

```console
(firedrake-disk-checkpointing) dham@IC-D46G33P392 deadlock % mpiexec -n 2 $VIRTUAL_ENV/bin/python adjoint_checkpoint_extruded_parallel_deadlock.py
 ▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▣▢▢▢▢ 184/200 [0:00:00]
Traceback (most recent call last):
  File "/Users/dham/src/firedrake-disk-checkpointing/lib/python3.10/site-packages/h5py/_hl/attrs.py", line 213, in create
Traceback (most recent call last):
  File "/Users/dham/src/firedrake-disk-checkpointing/lib/python3.10/site-packages/h5py/_hl/attrs.py", line 213, in create
    h5a.rename(self._id, self._e(tempname), self._e(name))
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
    h5a.rename(self._id, self._e(tempname), self._e(name))
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5a.pyx", line 121, in h5py.h5a.rename
  File "h5py/h5a.pyx", line 121, in h5py.h5a.rename
RuntimeError: Can't rename attribute (object header message is too large)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dham/src/firedrake-disk-checkpointing/src/checkpoint_tests/deadlock/adjoint_checkpoint_extruded_parallel_deadlock.py", line 33, in <module>
    outfile.save_function(T, name=f"AAAAAAAAAA_{timestep}")
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
RuntimeError: Can't rename attribute (object header message is too large)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dham/src/firedrake-disk-checkpointing/src/checkpoint_tests/deadlock/adjoint_checkpoint_extruded_parallel_deadlock.py", line 33, in <module>
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    outfile.save_function(T, name=f"AAAAAAAAAA_{timestep}")
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 716, in save_function
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 716, in save_function
    return self.save_function(g, idx=idx)
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 749, in save_function
    return self.save_function(g, idx=idx)
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 749, in save_function
    self.save_function(_f, idx=idx)
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 738, in save_function
    self._update_function_name_function_space_name_map(tmesh.name, mesh.name, {f.name(): V_name})
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1264, in _update_function_name_function_space_name_map
    self.save_function(_f, idx=idx)
  File "PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 738, in save_function
    self._update_function_name_function_space_name_map(tmesh.name, mesh.name, {f.name(): V_name})
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1264, in _update_function_name_function_space_name_map
    self._update_pickled_dict("function_name_function_space_name_map", new_item, tmesh_name, mesh_name)
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1242, in _update_pickled_dict
    self._update_pickled_dict("function_name_function_space_name_map", new_item, tmesh_name, mesh_name)
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1242, in _update_pickled_dict
    getattr(self, "_set_" + name)(*args, the_dict)
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1257, in _set_function_name_function_space_name_map
    getattr(self, "_set_" + name)(*args, the_dict)
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1257, in _set_function_name_function_space_name_map
    self._write_pickled_dict(path, PREFIX + "_function_name_function_space_name_map", new_item)
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1225, in _write_pickled_dict
    self._write_pickled_dict(path, PREFIX + "_function_name_function_space_name_map", new_item)
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1225, in _write_pickled_dict
    self.set_attr(path, name, self._pickle(the_dict))
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1315, in set_attr
    self.set_attr(path, name, self._pickle(the_dict))
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1315, in set_attr
    self.h5pyfile[path].attrs[key] = val
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
    self.h5pyfile[path].attrs[key] = val
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/Users/dham/src/firedrake-disk-checkpointing/lib/python3.10/site-packages/h5py/_hl/attrs.py", line 103, in __setitem__
  File "/Users/dham/src/firedrake-disk-checkpointing/lib/python3.10/site-packages/h5py/_hl/attrs.py", line 103, in __setitem__
    self.create(name, data=value)
  File "/Users/dham/src/firedrake-disk-checkpointing/lib/python3.10/site-packages/h5py/_hl/attrs.py", line 216, in create
    self.create(name, data=value)
  File "/Users/dham/src/firedrake-disk-checkpointing/lib/python3.10/site-packages/h5py/_hl/attrs.py", line 216, in create
    h5a.delete(self._id, self._e(tempname))
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5a.pyx", line 145, in h5py.h5a.delete
    h5a.delete(self._id, self._e(tempname))
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5a.pyx", line 145, in h5py.h5a.delete
KeyError: "Unable to delete attribute (can't locate attribute)"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dham/src/firedrake-disk-checkpointing/src/checkpoint_tests/deadlock/adjoint_checkpoint_extruded_parallel_deadlock.py", line 31, in <module>
    with CheckpointFile("foo.h5", "a") as outfile:
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 532, in __exit__
KeyError: "Unable to delete attribute (can't locate attribute)"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dham/src/firedrake-disk-checkpointing/src/checkpoint_tests/deadlock/adjoint_checkpoint_extruded_parallel_deadlock.py", line 31, in <module>
    with CheckpointFile("foo.h5", "a") as outfile:
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 532, in __exit__
    self.close()
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1340, in close
    self.close()
  File "/Users/dham/src/firedrake-disk-checkpointing/src/firedrake/firedrake/checkpointing.py", line 1340, in close
    self.viewer.destroy()
  File "PETSc/Viewer.pyx", line 113, in petsc4py.PETSc.Viewer.destroy
    self.viewer.destroy()
  File "PETSc/Viewer.pyx", line 113, in petsc4py.PETSc.Viewer.destroy
petsc4py.PETSc.Error: error code 76
[0] PetscViewerDestroy() at /Users/dham/src/firedrake-disk-checkpointing/src/petsc/src/sys/classes/viewer/interface/view.c:114
[0] PetscViewerDestroy_HDF5() at /Users/dham/src/firedrake-disk-checkpointing/src/petsc/src/sys/classes/viewer/impls/hdf5/hdf5v.c:100
[0] PetscViewerFileClose_HDF5() at /Users/dham/src/firedrake-disk-checkpointing/src/petsc/src/sys/classes/viewer/impls/hdf5/hdf5v.c:81
[0] Error in external library
[0] Error in HDF5 call H5Fclose() Status -1
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
petsc4py.PETSc.Error: error code 76
[1] PetscViewerDestroy() at /Users/dham/src/firedrake-disk-checkpointing/src/petsc/src/sys/classes/viewer/interface/view.c:114
[1] PetscViewerDestroy_HDF5() at /Users/dham/src/firedrake-disk-checkpointing/src/petsc/src/sys/classes/viewer/impls/hdf5/hdf5v.c:100
[1] PetscViewerFileClose_HDF5() at /Users/dham/src/firedrake-disk-checkpointing/src/petsc/src/sys/classes/viewer/impls/hdf5/hdf5v.c:81
[1] Error in external library
[1] Error in HDF5 call H5Fclose() Status -1
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
(firedrake-disk-checkpointing) dham@IC-D46G33P392 deadlock % 
```

**Comment by ksagiyam** [comment]
I was able to reproduce the crash with current PETSc fork of Firedrake even with the regular mesh (after 885 iterations). I have no idea, but will look into it.

**Comment by wence-** [comment]
I would be interested in output of `h5dump` for the crash that is shown. It might be that timestep names need to be saved in an actual data array rather than an attribute.

**Comment by ksagiyam** [comment]
`h5dump -A foo.h5`:
[h5dump_A_foo.txt](https://github.com/firedrakeproject/firedrake/files/9393687/h5dump_A_foo.txt)

`h5dump foo.h5`
[h5dump_foo.txt](https://github.com/firedrakeproject/firedrake/files/9393690/h5dump_foo.txt)

In practice I think we can use the timestep mode of `save_function`, but I think the above failing example indicates that `CheckpointFile` fails if we try to save more than 185 different functions or so.

**Comment by ksagiyam** [comment]
but indeed `function_name_function_space_name_map` might want to be saved as actual data; we had similar conversation before, but I decided to do otherwise, but the above example tells us that what I did was not a robust solution.

**Comment by dham** [comment]
This situation occurs in the case of adjoint checkpointing, where large numbers of Functions are saved but not really in a timestep sequence. I will give some thought to whether I can force them into such a sequence.

**Comment by dham** [comment]
Fixed in #2530
